File: scripts/src/tools/utils/sysinfo/bench/suites/thermal.py
import json
import os
import shutil
import subprocess
import sys
import time

from tools.core.process import capture as run
from tools.utils.sysinfo.bench.limits import SUSTAINED_SECONDS
from tools.utils.sysinfo.bench.record import HIB, HOST, LIB
from tools.utils.sysinfo.bench.suites import Job, MeasurementError, Output, tool_path, version_of
from tools.utils.sysinfo.formatting import as_dict

SAMPLE_INTERVAL = 2.0


def sensor_temperature():
    if sys.platform == "darwin":
        return None
    if not shutil.which("sensors"):
        return None
    result = run(["sensors", "-j"], timeout=10)
    if result.returncode != 0:
        return None
    try:
        payload = json.loads(result.stdout)
    except ValueError:
        return None
    best = None
    for chip, readings in payload.items():
        if not any(mark in chip.lower() for mark in ("k10temp", "coretemp", "zenpower")):
            continue
        for entries in as_dict(readings).values():
            for name, value in as_dict(entries).items():
                if "input" in name and isinstance(value, (int, float)):
                    best = value if best is None else max(best, value)
    return best


def package_clock():
    try:
        with open("/proc/cpuinfo", encoding="utf-8") as handle:
            values = [
                float(line.split(":", 1)[1])
                for line in handle.read().splitlines()
                if line.lower().startswith("cpu mhz")
            ]
    except OSError:
        return None
    return sum(values) / len(values) if values else None


def complaint(text):
    for line in reversed((text or "").splitlines()):
        if line.strip():
            return line.strip()
    return ""


def sustained(path, seconds):
    workers = str(max(1, os.cpu_count() or 2))
    process = subprocess.Popen(
        [path, "--matrix", workers, "--timeout", f"{seconds}s", "--metrics-brief"],
        stdout=subprocess.DEVNULL,
        stderr=subprocess.PIPE,
    )
    temperatures = []
    clocks = []
    deadline = time.monotonic() + seconds
    stopped_early = False
    try:
        while time.monotonic() < deadline:
            if process.poll() is not None:
                stopped_early = True
                break
            time.sleep(SAMPLE_INTERVAL)
            reading = sensor_temperature()
            if reading is not None:
                temperatures.append(reading)
            clock = package_clock()
            if clock is not None:
                clocks.append(clock)
    finally:
        if process.poll() is None:
            process.terminate()
        try:
            _output, refused = process.communicate(timeout=30)
        except subprocess.TimeoutExpired:
            # stress-ng ignores SIGTERM on some platforms.
            process.kill()
            _output, refused = process.communicate()
    # Popen returns before the child can fail to exec, so a stressor that rejects
    # its arguments still leaves poll() as None for the first pass. Without this
    # the single idle sample taken in that pass was published as the peak
    # temperature under sustained load, with no error.
    if stopped_early and process.returncode:
        raise MeasurementError(
            f"stress-ng exited {process.returncode}: {complaint(refused)}".rstrip(": ")
        )
    if not temperatures and not clocks:
        raise MeasurementError("no thermal telemetry was available during the load")
    values = {}
    if temperatures:
        values["thermal.peak"] = max(temperatures)
        tail = temperatures[len(temperatures) // 2 :]
        values["thermal.steady"] = sum(tail) / len(tail)
    if clocks:
        tail = clocks[len(clocks) // 2 :]
        values["cpu.sustained_clock"] = sum(tail) / len(tail)
    return values


def jobs(setting):
    seconds = SUSTAINED_SECONDS.get(setting.tier, 0)
    if not seconds:
        return []
    path = tool_path("stress-ng")
    if not path:
        return []
    return [
        Job(
            name="thermal",
            tool="stress-ng",
            version=version_of(path, args=("--version",), pattern=r"(\d[\d.]*)"),
            method="thermal/1.0.0",
            # Host scoped: these depend on the cooler and the case, not the OS.
            # Under platform scope two unrelated machines running the same
            # distro had their temperatures compared and labelled better/worse.
            outputs=(
                Output("thermal.peak", "°C", LIB, HOST),
                Output("thermal.steady", "°C", LIB, HOST),
                Output("cpu.sustained_clock", "MHz", HIB, HOST),
            ),
            measure=lambda: sustained(path, seconds),
            repeat=False,
            detail={"seconds": seconds, "stressor": "matrix", "interval": SAMPLE_INTERVAL},
        )
    ]

File: scripts/rust/crates/sysinfo/src/bench/suites/thermal.rs
use super::{Job, Measurement, capture_for, job, output, tool_path, version};
use crate::bench::runner::Setting;
use serde_json::{Value, json};
use std::{
    fs,
    path::Path,
    process::Command,
    sync::mpsc,
    time::{Duration, Instant},
};
pub fn sensor_temperature(timeout: Duration) -> Option<f64> {
    if cfg!(target_os = "macos") {
        return None;
    }
    let result = capture_for(
        Command::new("sensors").arg("-j"),
        timeout.min(Duration::from_secs(10)),
    )
    .ok()?;
    if !result.status.success() {
        return None;
    }
    let payload: Value = serde_json::from_slice(&result.stdout).ok()?;
    let mut best: Option<f64> = None;
    for (chip, readings) in payload.as_object()? {
        if !["k10temp", "coretemp", "zenpower"]
            .iter()
            .any(|mark| chip.to_lowercase().contains(mark))
        {
            continue;
        }
        for entries in readings
            .as_object()
            .into_iter()
            .flat_map(|object| object.values())
        {
            for (name, value) in entries.as_object().into_iter().flatten() {
                if name.contains("input")
                    && let Some(value) = value.as_f64()
                {
                    best = Some(best.map_or(value, |before| before.max(value)));
                }
            }
        }
    }
    best
}
pub fn package_clock() -> Option<f64> {
    let text = fs::read_to_string("/proc/cpuinfo").ok()?;
    let values = text
        .lines()
        .filter(|line| line.to_lowercase().starts_with("cpu mhz"))
        .filter_map(|line| line.split_once(':')?.1.trim().parse::<f64>().ok())
        .collect::<Vec<_>>();
    (!values.is_empty()).then(|| values.iter().sum::<f64>() / values.len() as f64)
}
fn sample_before_deadline<T>(
    deadline: Instant,
    probe: impl FnOnce(Duration) -> Option<T>,
) -> Option<T> {
    let remaining = deadline.checked_duration_since(Instant::now())?;
    let value = probe(remaining)?;
    (Instant::now() < deadline).then_some(value)
}
pub fn sustained(path: &Path, seconds: u64) -> Result<Measurement, String> {
    let path = path.to_path_buf();
    let workers = std::thread::available_parallelism().map_or(2, usize::from);
    let (sender, receiver) = mpsc::channel();
    let deadline = Instant::now() + Duration::from_secs(seconds);
    let worker = std::thread::spawn(move || {
        let reply = hostkit::process::output_terminating(
            Command::new(path).args([
                "--matrix",
                &workers.to_string(),
                "--timeout",
                &format!("{seconds}s"),
                "--metrics-brief",
            ]),
            hostkit::process::CaptureLimits {
                stdout: 0,
                stderr: 128 * 1024,
            },
            Duration::from_secs(seconds),
            Duration::from_secs(30),
            &ui_terminal::termination_requested,
        )
        .map_err(|error| error.to_string());
        let _ = sender.send(reply);
    });
    let mut temperatures = Vec::new();
    let mut clocks = Vec::new();
    let result = loop {
        match receiver.recv_timeout(Duration::from_secs(2)) {
            Ok(result) => break result,
            Err(mpsc::RecvTimeoutError::Timeout) => {
                if let Some(value) = sample_before_deadline(deadline, sensor_temperature) {
                    temperatures.push(value);
                }
                if let Some(value) = sample_before_deadline(deadline, |_| package_clock()) {
                    clocks.push(value);
                }
            }
            Err(error) => break Err(error.to_string()),
        }
    };
    worker.join().map_err(|_| "thermal worker failed")?;
    let result = result?;
    if !result.deadline_reached && !result.output.status.success() {
        let text = String::from_utf8_lossy(&result.output.stderr);
        let detail = text
            .lines()
            .rev()
            .find(|line| !line.trim().is_empty())
            .unwrap_or("");
        return Err(format!(
            "stress-ng exited {}: {detail}",
            result.output.status
        ));
    }
    if temperatures.is_empty() && clocks.is_empty() {
        return Err("no thermal telemetry was available during the load".into());
    }
    let mut values = Vec::new();
    if !temperatures.is_empty() {
        values.push((
            "thermal.peak".into(),
            vec![
                temperatures
                    .iter()
                    .copied()
                    .fold(f64::NEG_INFINITY, f64::max),
            ],
        ));
        let tail = &temperatures[temperatures.len() / 2..];
        values.push((
            "thermal.steady".into(),
            vec![tail.iter().sum::<f64>() / tail.len() as f64],
        ));
    }
    if !clocks.is_empty() {
        let tail = &clocks[clocks.len() / 2..];
        values.push((
            "cpu.sustained_clock".into(),
            vec![tail.iter().sum::<f64>() / tail.len() as f64],
        ));
    }
    Ok(Measurement::values(values))
}

pub fn jobs(setting: &Setting) -> Result<Vec<Job>, String> {
    if !setting.accepts("thermal") || setting.tier == "quick" {
        return Ok(Vec::new());
    }
    let Some(path) = tool_path(&["stress-ng"]) else {
        return Ok(Vec::new());
    };
    let seconds = if setting.tier == "heavy" { 120 } else { 60 };
    let ver = version(&path, &["--version"], r"(\d[\d.]*)");
    let mut result = job(
        "thermal",
        "stress-ng",
        &ver,
        "thermal/1.0.0",
        vec![
            output("thermal.peak", "°C", "LIB", "host"),
            output("thermal.steady", "°C", "LIB", "host"),
            output("cpu.sustained_clock", "MHz", "HIB", "host"),
        ],
        json!({"seconds":seconds,"stressor":"matrix","interval":2.0}),
        move || sustained(&path, seconds),
    );
    result.repeat = false;
    Ok(vec![result])
}

#[cfg(test)]
#[path = "../../../tests/thermal_tests.rs"]
mod tests;
